tools/checked-c-convert/ProgramInfo.h

```diff
@@ -33,80 +33,123 @@
+
 private:
-  GlobalSymbolKind Kind;
-  std::string Name;
-  PersistentSourceLoc Loc;
+  ConstraintVariableKind Kind;
+protected:
+  std::string BaseType;
 public:
-  GlobalSymbol(GlobalSymbolKind K, std::string N, PersistentSourceLoc P) 
-    : Kind(K),Name(N),Loc(P) {} 
-  GlobalSymbolKind getKind() const { return Kind; }
+  ConstraintVariable(ConstraintVariableKind K, std::string T) : 
+    Kind(K),BaseType(T) {}
 
-  bool operator<(const GlobalSymbol &O) const {
-    return Loc < O.Loc;
-  }
+  virtual std::string mkString(Constraints::EnvironmentMap &E) = 0;
+  virtual void print(llvm::raw_ostream &O) const = 0;
+  virtual void dump() const = 0;
+  virtual void constrainTo(Constraints &CS, ConstAtom *C) = 0;
+  virtual bool anyChanges(Constraints::EnvironmentMap &E) = 0;
 
-  std::string getName() { return Name; }
+  std::string getTy() { return BaseType; }
 };
 
-class GlobalVariableSymbol : public GlobalSymbol {
+class PointerVariableConstraint;
+class FunctionVariableConstraint;
+
+// Represents an individual constraint on a pointer variable. 
+// This could contain a reference to a FunctionVariableConstraint
+// in the case of a function pointer declaration.
+class PointerVariableConstraint : public ConstraintVariable {
 private:
-  std::set<uint32_t> ConstraintVars;
+  CVars vars;
+  FunctionVariableConstraint *FV;
 public:
-  GlobalVariableSymbol(std::string N, PersistentSourceLoc P, 
-    std::set<uint32_t> S) 
-    : GlobalSymbol(Variable, N, P),ConstraintVars(S) {}
+  PointerVariableConstraint(CVars V, std::string T) : 
+    ConstraintVariable(PointerVariable, T),vars(V),FV(nullptr) {}
+  PointerVariableConstraint(clang::DeclaratorDecl *D, uint32_t &K,
+    Constraints &CS);
+  PointerVariableConstraint(const clang::Type *Ty, uint32_t &K,
+    std::string N, Constraints &CS);
 
-  std::set<uint32_t> &getVars() {
-    return ConstraintVars;
-  }
+  const CVars &getCvars() const { return vars; }
 
-  static bool classof(const GlobalSymbol *S) {
-    return S->getKind() == Variable;
+  static bool classof(const ConstraintVariable *S) {
+    return S->getKind() == PointerVariable;
   }
+
+  std::string mkString(Constraints::EnvironmentMap &E);
+
+  FunctionVariableConstraint *getFV() { return FV; }
+
+  void print(llvm::raw_ostream &O) const ;
+  void dump() const { print(llvm::errs()); }
+  void constrainTo(Constraints &CS, ConstAtom *C);
+  bool anyChanges(Constraints::EnvironmentMap &E);
 };
 
-class GlobalFunctionSymbol : public GlobalSymbol {
+typedef PointerVariableConstraint PVConstraint;
+
+// Constraints on a function type. Also contains a 'name' parameter for 
+// when a re-write of a function pointer is needed.
+class FunctionVariableConstraint : public ConstraintVariable {
 private:
-  std::vector<std::set<uint32_t> > ParameterConstraintVars;
-  std::set<uint32_t> ReturnConstraintVars;
+  std::set<ConstraintVariable*> returnVars;
+  std::vector<std::set<ConstraintVariable*>> paramVars;
```

Comment:
Please add comments explaining how these constraints work.  As far as I can tell, the first one is N constraints on the return value.   The second is a vector of N constrains, where each element k of the vector corresponds to the kth parameter.